zonemaster/zonemaster-backend
# Separate config section for each db engine

## Purpose

This PR is about having separate a separate config section for each db engine.


## Context

Fixes #719.

As @matsduf pointed out, PR #717 must be merged before this is merged or else there is a risk of filling up the log files.


## Changes

Config format:
* New sections are added. The new section names are upper-case-only for consistency with previous sections.
* The old properties are deprecated.

Implementation:
* New accessor methods for the properties are added to Config.pm. They're named after the config section and property name, with an underscore in between. There is some precedence for this naming scheme in Config.pm.
* If the new properties are absent in the config file, the old ones are used as fallbacks to make sure old config files keep working.
* The accessors for the old properties are removed.
* The DB_connection_string() method didn't really belong in the Config module so I inlined it into the DB modules.

Installation:
* Newly obsolete steps in the installation instruction are cleaned up.
* The included config file is updated to the new format.

CI:
* The backend config files for travis are updated to the new format.

Database initialization:
* The initialization scripts now use the adapter class - they used to read accessors on the config object and construct its own handle. (Note that the SQLite initialization script is only used in the installation instruction while the MySQL and PostgreSQL dittos are only used for CI.)

Database upgrade:
* The database upgrade scripts are updated the same way as the database initialization scripts.

Logging:
* It's now possible to override ZM_BACKEND_RPCAPI_LOGLEVEL and ZM_BACKEND_TESTAGENT_LOGLEVEL when starting the daemons using the supplied init scripts.
* Logging of database connections is now more informative.
* Log messages emitted during the loading for the config file during startup is now output to the terminal (instead of being silently ignored).

## How to test this PR

This is a new feature, so older versions of Zonemaster need not be tested.

For more convenient testing of the database settings, set ZM_BACKEND_RPCAPI_LOGLEVEL and ZM_BACKEND_TESTAGENT_LOGLEVEL to notice or lower.

For each database engine:
1. Get the source from this PR.
   ```sh
   git clone --origin mattias-p --branch 719-db-sections https://github.com/mattias-p/zonemaster-backend.git
   cd zonemaster-backend
   ```
2. Verify that both daemons can be installed and that the SQLite database initialization works.
   1. Install Backend for according to the installation instruction, but instead of `cpanm Zonemaster::Backend` run `cpanm .`.
3. Verify that the installation uses the new configuration format.
      ```sh
      cat /etc/zonemaster/backend_config.ini
      ```
4. Verify that both daemons can be started using the old configuration format.
   1. Copy the share/backend_config.ini from the current develop branch into /etc/zonemaster.
      ```sh
      git remote add origin https://github.com/zonemaster/zonemaster-backend.git
      git fetch origin
      git checkout origin/develop
      sudo install -v -m 640 -g zonemaster share/backend_config.ini /etc/zonemaster/
      ```
   2. Update the DB.engine property for the current database engine according to the installation instruction.
   3. Stop the daemons.
      ```sh
      sudo service stop zm-rpcapi
      sudo sercive status zm-rpcapi
      sudo service stop zm-testagent
      sudo sercive status zm-testagent
      ```
   4. Start the daemons.
      ```sh
      sudo service start zm-rpcapi
      sudo sercive status zm-rpcapi
      sudo service start zm-testagent
      sudo sercive status zm-testagent
      ```
5. Verify that deprecation warnings are emitted to the log files of both rpcapi and testagent if the old configurations properties are present when the configuration file is loaded.
   ```sh
   cat /var/log/zonemaster/zm-testagent.log | grep deprecated
   journalctl -u zm-rpcapi.service | grep deprecated
   ```
